src/Workspaces/Core/Portable/FindSymbols/FindReferences/MetadataUnifyingEquivalenceComparer.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal sealed class MetadataUnifyingEquivalenceComparer : IEqualityComparer<ISymbol>
+    {
+        public static readonly IEqualityComparer<ISymbol> Instance = new MetadataUnifyingEquivalenceComparer();
+
+        private MetadataUnifyingEquivalenceComparer()
+        {
+        }
+
+        public bool Equals(ISymbol x, ISymbol y)
+        {
+            // If either symbol is from source, then we must do stricter equality. Consider this:
+            //
+            //     S1 <-> M <-> S2     (where S# = source symbol, M = some metadata symbol)
+            //
+            // In this case, imagine that both the comparisons denoted by <-> were done with the
+            // SymbolEquivalenceComparer, and returned true. If S1 and S2 were from different projects,
+            // they would compare false but transitivity would say they must be true. Another way to think
+            // of this is any use of a source symbol "poisons" the comparison and requires it to be stricter.
+            if (x == null || y == null || IsInSource(x) || IsInSource(y))
```

Comment:
This is why I made the change to FindSourceSymbolAsync. Before I made this change, we had exactly the issue as you described it: a source symbol from P1, and that same source symbol being wrapped for retargeting in P2. Before adding symbols to the overall symbol list, we do call FindSourceSymbolAsync.